module/os/linux/zfs/arc_os.c

```diff
@@ -268,6 +268,24 @@ __arc_shrinker_func(struct shrinker *shrink, struct shrink_control *sc)
 	 */
 	if (pages > 0) {
 		arc_reduce_target_size(ptob(sc->nr_to_scan));
+
+		/*
+		 * Repeated calls to the arc shrinker can reduce arc_c
+		 * drastically, potentially all the way to arc_c_min.  While
+		 * arc_c is below arc_size, ZFS can't process read/write
+		 * requests, because arc_get_data_impl() will block.  To
+		 * ensure that arc_c doesn't shrink faster than the adjust
+		 * thread can keep up, we wait for eviction here.
+		 */
+		mutex_enter(&arc_adjust_lock);
+		if (arc_is_overflowing()) {
+			arc_adjust_needed = B_TRUE;
+			zthr_wakeup(arc_adjust_zthr);
```

Comment:
Now that we're doing this, do we still need the call to:
```
		/*
		 * We've shrunk what we can, wake up threads.
		 */
		cv_broadcast(&arc_adjust_waiters_cv);
```
a few lines below...

it looks like we'll already call `cv_broadcast` in `arc_adjust_cb`, so perhaps we don't need to explicitly call it in this function?

Reply:
Dropping it does look safe to me.  However, I wonder if we shouldn't instead convert it to a `cv_signal()` and apply the same logic described in the `arc_evict_state_impl()` comment for the non-overflow case here (i.e. when we've only reaped the kmem caches).

```c
                        /*
                         * If arc_size isn't overflowing, signal any
                         * threads that might happen to be waiting.
                         *
                         * For each header evicted, we wake up a single
                         * thread. If we used cv_broadcast, we could
                         * wake up "too many" threads causing arc_size
                         * to significantly overflow arc_c; since
                         * arc_get_data_impl() doesn't check for overflow
                         * when it's woken up (it doesn't because it's
                         * possible for the ARC to be overflowing while
                         * full of un-evictable buffers, and the
                         * function should proceed in this case).
                         *
                         * If threads are left sleeping, due to not
                         * using cv_broadcast here, they will be woken
                         * up via cv_broadcast in arc_adjust_cb() just
                         * before arc_adjust_zthr sleeps.
                         */
```

Reply:
The `cv_broadcast(&arc_adjust_waiters_cv);` has been wrong since 3ec34e5, because the arc shrinker doesn't evict anything.  This CV is used to indicate that the arc has been "adjusted", i.e. we've evicted from it.  It's correctly signalled in `arc_evict_state_impl()`, and there's no need to do so here (because we don't evict anything -- now we wait for arc_evict_state_impl() to evict, and it handles signaling the CV).  This seems very safe to remove, so I'm fine with doing it as part of this commit if you like.